src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DecisionDagRewriter.cs

```diff
@@ -410,53 +412,510 @@ private void GenerateTest(BoundExpression test, BoundDecisionDagNode whenTrue, B
             private bool GenerateSwitchDispatch(BoundDecisionDagNode node, HashSet<BoundDecisionDagNode> loweredNodes)
             {
                 Debug.Assert(!loweredNodes.Contains(node));
-
-                // We only generate a switch dispatch if we have two or more value tests in a row
-                if (!(node is BoundTestDecisionDagNode firstTestNode &&
-                     firstTestNode.Test is BoundDagValueTest firstTest &&
-                     firstTestNode.WhenFalse is BoundTestDecisionDagNode whenFalse &&
-                     whenFalse.Test is BoundDagValueTest secondTest &&
-                     !loweredNodes.Contains(whenFalse) &&
-                     !this._dagNodeLabels.ContainsKey(whenFalse) &&
-                     firstTest.Input.Equals(secondTest.Input) &&
-                     firstTest.Input.Type.IsValidV6SwitchGoverningType()))
-                {
-                    // https://github.com/dotnet/roslyn/issues/12509 Could optimize float, double, decimal value switches. For now use if-then-else
+                if (!CanGenerateSwitchDispatch(node, loweredNodes))
                     return false;
+
+                var input = ((BoundTestDecisionDagNode)node).Test.Input;
+                ValueDispatchNode n = GatherValueDispatchNodes(node, loweredNodes, input);
+                LowerValueDispatchNode(n, input);
+                return true;
+            }
+
+            private bool CanGenerateSwitchDispatch(BoundDecisionDagNode node, HashSet<BoundDecisionDagNode> loweredNodes)
+            {
+                switch (node)
+                {
+                    // These are the forms worth optimizing.
+                    case BoundTestDecisionDagNode { WhenFalse: BoundTestDecisionDagNode test2 } test1:
+                        return CanDispatch(test1, test2);
+                    case BoundTestDecisionDagNode { WhenTrue: BoundTestDecisionDagNode test2 } test1:
+                        return CanDispatch(test1, test2);
+                    default:
+                        // Other cases are just as well done with a single test.
+                        return false;
+                }
+
+                bool CanDispatch(BoundTestDecisionDagNode test1, BoundTestDecisionDagNode test2)
+                {
+                    if (loweredNodes.Contains(test2))
```

Comment:
>if (loweredNodes.Contains(test2)) [](start = 20, length = 33)

Can we remove this `Contains()` call from the RELEASE build, by asserting after the later call? For instance:
```C#
if (this._dagNodeLabels.ContainsKey(test2))
   return false;
Debug.Assert(!loweredNodes.Contains(test2));
```

Reply:
Yes; good idea.

---
In reply to: [390484650](https://github.com/dotnet/roslyn/pull/42313#discussion_r390484650) [](ancestors = 390484650)